src/adapters/morpho-aave-v2/common/morphoBasePoolAdapter.ts

```diff
@@ -0,0 +1,633 @@
+  MorphoAaveV2__factory,
+  MorphoAToken__factory,
+  MorphoAaveV2Lens__factory,
+} from '../contracts'
+import { TransferEvent } from '../contracts/MorphoAToken'
+
+type MorphoAaveV2PeerToPoolAdapterMetadata = Record<
+  string,
+  {
+    protocolToken: Erc20Metadata
+    underlyingToken: Erc20Metadata
+  }
+>
+
+const morphoAaveV2ContractAddresses: Partial<
+  Record<Protocol, Partial<Record<Chain, string>>>
+> = {
+  [Protocol.MorphoAaveV2]: {
+    [Chain.Ethereum]: '0x777777c9898d384f785ee44acfe945efdff5f3e0',
+  },
+}
+
+export abstract class MorphoBasePoolAdapter implements IMetadataBuilder {
+  protocolId: Protocol
+  chainId: Chain
+
+  protected provider: CustomJsonRpcProvider
+
+  constructor({
+    provider,
+    chainId,
+    protocolId,
+    adaptersController,
+  }: ProtocolAdapterParams) {
+    this.provider = provider
+    this.chainId = chainId
+    this.protocolId = protocolId
+    this.adaptersController = adaptersController
+  }
+
+  lensAddress = '0x507fa343d0a90786d86c7cd885f5c49263a91ff4'
+
+  adaptersController: AdaptersController
+
+  abstract getProtocolDetails(): ProtocolDetails
+
+  async buildMetadata() {
+    const morphoAaveV2Contract = MorphoAaveV2__factory.connect(
+      morphoAaveV2ContractAddresses[this.protocolId]![this.chainId]!,
+      this.provider,
+    )
+
+    const metadataObject: MorphoAaveV2PeerToPoolAdapterMetadata = {}
+
+    const markets = await morphoAaveV2Contract.getMarketsCreated()
+
+    const promises = markets.map(async (marketAddress) => {
+      const aTokenContract = MorphoAToken__factory.connect(
+        marketAddress,
+        this.provider,
+      )
+
+      const supplyTokenAddress = await aTokenContract
+        .UNDERLYING_ASSET_ADDRESS()
+        .catch((err) => {
+          if (err) return ZERO_ADDRESS
+          throw err
+        })
+
+      const underlyingTokenPromise = getTokenMetadata(
+        supplyTokenAddress,
+        this.chainId,
+        this.provider,
+      )
+      const protocolTokenPromise = getTokenMetadata(
+        marketAddress,
+        this.chainId,
+        this.provider,
+      )
+      const [protocolToken, underlyingToken] = await Promise.all([
+        protocolTokenPromise,
+        underlyingTokenPromise,
+      ])
+
+      metadataObject[protocolToken.address] = {
+        protocolToken,
+        underlyingToken,
+      }
+    })
+
+    await Promise.all(promises)
+
+    return metadataObject
+  }
+
+  async getProtocolTokens(): Promise<Erc20Metadata[]> {
+    return Object.values(await this.buildMetadata()).map(
+      ({ protocolToken }) => protocolToken,
+    )
+  }
+
+  protected async fetchProtocolTokenMetadata(
+    protocolTokenAddress: string,
+  ): Promise<Erc20Metadata> {
+    const { protocolToken } = await this.fetchPoolMetadata(protocolTokenAddress)
+
+    return protocolToken
+  }
+
+  private async fetchPoolMetadata(protocolTokenAddress: string) {
+    const poolMetadata = (await this.buildMetadata())[protocolTokenAddress]
```

Comment:
it's rebuilding all the metadata for only one accessed value, its really not efficient